source/slang/slang-check-decl.cpp

```diff
@@ -8475,19 +8829,327 @@ namespace Slang
             }
             if (seqStmtChild->stmts.getCount() == 0)
                 continue;
-            seqStmt->stmts.insert(insertOffset, seqStmtChild);
+
+            auto seqStmt = _ensureCtorBodyIsSeqStmt(m_astBuilder, ctor);
+            seqStmt->stmts.insert(ctorInfo.m_insertOffset++, seqStmtChild);
         }
 
-        if (structDeclInfo.defaultCtor)
+        // Note: we assume only 1 inheritance decl currently.
+        // Pre-calculate if we have a base-type and its associated ctor-list
+        // We also must ensure inheritance-decl is checked, else we may not have up-to-date 'DerivativeMemberAttribute' modifiers
+        DeclAndCtorInfo* baseStructInfo = nullptr;
+        for(auto& i : inheritanceInfoList)
         {
-            auto seqStmt = as<SeqStmt>(as<BlockStmt>(structDeclInfo.defaultCtor->body)->body);
-            if (seqStmt && seqStmt->stmts.getCount() == 0)
+            ensureDecl(i.m_inheritanceDecl, DeclCheckState::VarInitExprAreChecked);
+            if(as<StructDecl>(i.m_inheritanceBaseDecl))
+                baseStructInfo = &i;
+        }
+
+        // pre-calculate any requirements of a CudaHostAttribute
+        HashSet<VarDeclBase*> requiresCudaHostModifier;
+        for (auto member : membersOfStructDeclInstance)
+            if (containsTargetType<TorchTensorType>(m_astBuilder, member.getDecl()->type.type))
```

Comment:
This is going to be bad for performance because it revisits the entire type hierarchy for all the member types.

Reply:
An alternative is to add the required CudaHost attribute in an IR pass to make things cleaner.